src/view/region.rs

```diff
@@ -1,121 +1,306 @@
+    pub fn new(graph: G, hierarchy: &'g Hierarchy, root: NodeIndex) -> Self {
         let mut is_descendant = HashMap::new();
         is_descendant.insert(root, false);
         Self {
+            graph,
+            region_root: root,
             hierarchy,
-            root,
-            is_descendant: RefCell::new(is_descendant),
+            is_descendant: RwLock::new(is_descendant),
         }
     }
 
     /// Check if a node is a descendant of the root node.
     ///
     /// Caches the result of the check.
     pub fn is_descendant(&self, node: NodeIndex) -> bool {
-        if let Some(is_descendant) = self.is_descendant.borrow().get(&node) {
-            // We have already checked this node.
-            return *is_descendant;
+        if node == self.region_root {
+            return true;
         }
-        // Traverse the ancestors until we see a node we have already checked.
-        let cache = self.is_descendant.borrow();
-        let ancestors: Vec<_> = iter::successors(Some(node), |node| {
-            let parent = self.hierarchy.parent(*node)?;
-            match cache.contains_key(&parent) {
-                true => None,
-                false => Some(parent),
+
+        // First, access the cache read-only to check if the node has already been visited.
+        // And compute whether the node is a descendant otherwise.
+        let (ancestors, is_descendant) = {
+            let cache = self
+                .is_descendant
+                .read()
+                .expect("The Region cache is poisoned.");
+
+            if let Some(is_descendant) = cache.get(&node) {
+                // We have already checked this node.
+                return *is_descendant;
             }
-        })
-        .collect();
-        let first_visited_ancestor = self.hierarchy.parent(*ancestors.last().unwrap());
-        let is_descendant = first_visited_ancestor == Some(self.root)
-            || first_visited_ancestor
-                .map_or(false, |ancestor| cache.get(&ancestor).copied().unwrap());
-        drop(cache);
-
-        // Update the cache for all the unvisited ancestors.
-        let mut cache_mut = self.is_descendant.borrow_mut();
+            // Traverse the ancestors until we see a node we have already checked, or the root.
+            let ancestors: Vec<_> = iter::successors(Some(node), |node| {
+                let parent = self.hierarchy.parent(*node)?;
+                match cache.contains_key(&parent) {
+                    true => None,
+                    false => Some(parent),
+                }
+            })
+            .collect();
+            let first_visited_ancestor = self.hierarchy.parent(*ancestors.last().unwrap());
+            let is_descendant = first_visited_ancestor == Some(self.region_root)
+                || first_visited_ancestor
+                    .map_or(false, |ancestor| cache.get(&ancestor).copied().unwrap());
+            drop(cache);
```

Comment:
OOI why does this need an explicit drop? (could be worth a comment)

Reply:
Yes, it's not needed but I'd rather be explicit when dealing with locks.
I added a comment explaining it.